### DISCUSSION TITLE
Microsoft.ClearScript.Windows.Core VBScriptEngine - initializing.

### DISCUSSION THREAD

**Discussion by inavnapro** [ClearFoundry/ClearScript#347] [Q&A] [answered]
Hi All,

How to initialize VBScriptEngine() in windows core?  The constructor takes argument Isynclnvoker.  How do I initialize the IsyncInvoker when creating vb script instance .NET core.  For example:

/* Code example */
private VBScriptEngine vbsinstace;

public VBScriptEngine GetVBSCriptEngine()
{
    if(vbsintance == null || vbsinstace.CheckAccess() == false)
    {
          vbsinstance = new VBScriptEngine(---NEED HELP WITH THIS PARAM---)
    }
   return vbsinstance;
}

Please advise.  Thanks!

**Comment by ClearScriptLib** [comment, marked as answer]
Hi @inavnapro,

Like all Windows Script engines, VBScript has _thread affinity_, meaning that each instance can only be used on the thread that created it. Any attempt to use the script engine on the wrong thread results in an exception or possibly a crash.

ClearScript provides two managed wrappers for VBScript:

1. [`Microsoft.ClearScript.Windows.VBScriptEngine`](https://clearscript.clearfoundry.net/Reference/html/T_Microsoft_ClearScript_Windows_VBScriptEngine.htm).
2. [`Microsoft.ClearScript.Windows.Core.VBScriptEngine`](https://clearscript.clearfoundry.net/Reference/html/T_Microsoft_ClearScript_Windows_Core_VBScriptEngine.htm).

The first class doesn't require an external implementation of [`ISyncInvoker`](https://clearscript.clearfoundry.net/Reference/html/T_Microsoft_ClearScript_Windows_Core_ISyncInvoker.htm). Instead, it provides its own implementation backed by an instance of [`Dispatcher`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.threading.dispatcher). This version of `VBScriptEngine` is intended for desktop or console applications that have a main thread (or one or more UI threads). A multithreaded server could use it, of course, but it would probably have to manage a pool of dedicated script execution threads.

The other class allows the host to enforce thread affinity externally by implementing `ISyncInvoker`. This is mainly intended for multithreaded server use, but it could also be a good fit for simple single-threaded applications. If the host can ensure that the script engine is always invoked on the correct thread, it can use [`NullSyncInvoker.Instance`](https://clearscript.clearfoundry.net/Reference/html/F_Microsoft_ClearScript_Windows_Core_NullSyncInvoker_Instance.htm) to bypass thread affinity enforcement altogether.

Good luck!

**Reply by inavnapro** [reply]
Thanks!
